adding gitleaks precommit config

gfreeman-navapbc opened the pull request:
## 🎫 Ticket

https://jira.cms.gov/browse/PLT-3 Add pre-commit config for gitleaks to all repos

## 🛠 Changes
Add pre-commit config for gitleaks to all repos

kyeah commented on README.md:
```diff
@@ -256,3 +257,27 @@ make load-fixtures
 Follow installing go + vscode [setup guide](https://marketplace.visualstudio.com/items?itemName=golang.go#getting-started).
 Additional settings found under `.vscode/settings.json` allow tests to be run within vscode.
+  
+# Installing and Using Pre-commit
```
can we move this to the top (under the "Dependencies" section) and/or link to the section from the Dependencies list?

kyeah replied:
same comment for the other repos — if we could keep this near the top so people don't miss it and forget to use pre-commit, that'll be helpful.

kyeah reviewed (approved):
config looks good to me and instructions worked on my Mac. Would prefer to move the documentation up instead of having it sit at the bottom of the README in these repos. thanks!

gsf commented on README.md:
```diff
@@ -25,6 +26,31 @@ The files committed in the `shared_files/encrypted` directory hold secret inform
+
 #### Git hook
 
 To avoid committing and pushing unencrypted secret files, use the included `scripts/pre-commit` git pre-commit hook by running the following script from the repository root directory:
```
Maybe we want to remove this if pre-commit serves the same purpose.

gsf commented on README.md:
````diff
@@ -25,9 +26,30 @@ The files committed in the `shared_files/encrypted` directory hold secret inform
+This will download and install the pre-commit hooks specified in `.pre-commit-config.yaml`.
 
 ```
 cp ops/pre-commit .git/hooks
````
We could also add `go-imports` from https://github.com/TekWizely/pre-commit-golang to the `.pre-commit-config.yaml` to maintain the previous hook functionality but I'd want @kyeah's opinion on that.

kyeah replied:
Oh that would be nice to move everything to the pre-commit.config.yaml @gfreeman-navapbc @gsf. We would definitely want to maintain the previous functionality. I assume the gitleaks config would handle the ansible vault secrets check but would want to confirm as well:

https://github.com/CMSgov/bcda-app/blob/2ab83d60a3bb007661a7fd2fad23c547096a9e3c/ops/pre-commit#L16-L29

cc. @laurenkrugen-navapbc I think we had created a ticket to improve the pre-commit, this may make it easier to do so in the future since there's a lot of built-in utilities in https://github.com/TekWizely/pre-commit-golang.

gsf commented on README.md:
```diff
@@ -25,15 +26,30 @@ The files committed in the `shared_files/encrypted` directory hold secret inform
+
+##### Step 2: Install the hooks
+
+Run the following command to install the gitleaks hook:
```
Remove gitleaks here as it now includes the go-imports

```suggestion
Run the following command to install the hook:
```

gsf commented on README.md:
````diff
@@ -25,15 +26,30 @@ The files committed in the `shared_files/encrypted` directory hold secret inform
 ```
 
-The pre-commit hook will also ensure that any added, copied, or modified go files are formatted properly.
+This will download and install the pre-commit hooks specified in `.pre-commit-config.yaml`.
````
Maybe keep some detail from the deleted line like

```suggestion
This will download and install the pre-commit hooks specified in `.pre-commit-config.yaml`, which includes gitleaks for secret scanning and go-imports to ensure that any added, copied, or modified go files are formatted properly.
```

kyeah commented:
Thanks, looks good overall!

Could we delete the old pre-commit file? I also noticed that somehow the old pre-commit is still showing error messages and being run as part of the pre-commit. Is that expected? Even when I delete the .git/hooks/pre-commit file and `ops/pre-commit` and run `pre-commit install`, the old secrets check is still there 🤔 

```
bcda-app: ./.git/hooks/pre-commit
attempting to commit an unencrypted secret: shared_files/encrypted/api.yml; aborting
... <output from new pre-commit>
```

The additional output looks good though from gitleaks and such:

```
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /Users/kyeah/.cache/pre-commit/patch1684354882-69137.
Detect hardcoded secrets.................................................Failed
- hook id: gitleaks
- exit code: 1

○
    │╲
    │ ○
    ○ ░
    ░    gitleaks

Finding:     REDACTED
Secret:      REDACTED
RuleID:      private-key
Entropy:     6.033833
File:        shared_files/encrypted/bfd-dev-test-key.pem
Line:        1
Fingerprint: shared_files/encrypted/bfd-dev-test-key.pem:private-key:1

Finding:     BCDA_SSAS_SECRET=REDACTED
BLACKLIST_CLIENT_ID...
Secret:      REDACTED
RuleID:      generic-api-key
Entropy:     3.862885
File:        shared_files/encrypted/local.env
Line:        16
Fingerprint: shared_files/encrypted/local.env:generic-api-key:16

Finding:     BLACKLIST_CLIENT_ID=REDACTED
BLACKLIST_CLIENT_SE...
Secret:      REDACTED
RuleID:      generic-api-key
Entropy:     3.708003
File:        shared_files/encrypted/local.env
Line:        17
Fingerprint: shared_files/encrypted/local.env:generic-api-key:17

Finding:     BLACKLIST_CLIENT_SECRET=REDACTED
CCLF_CUTOFF_DATE_DA...
Secret:      REDACTED
RuleID:      generic-api-key
Entropy:     3.843254
File:        shared_files/encrypted/local.env
Line:        18
Fingerprint: shared_files/encrypted/local.env:generic-api-key:18

4:21PM INF 1 commits scanned.
4:21PM INF scan completed in 67.6ms
4:21PM WRN leaks found: 4
```

kyeah commented:
@gfreeman-navapbc A couple notes from testing the go-import functionality:

1. it requires a manual install:

```
error: command not found: goimports
```

I had to run `go install golang.org/x/tools/cmd/goimports@latest`. Could we document that? 

We could alternatively just rely on the basic `go-fmt` functionality that comes with Go, but if goimports provides some extra tooling I'd prefer to use it I think. It seems to come directly from the Golang team so I don't mind requiring it. cc. @laurenkrugen-navapbc @karinamzalez 

2. It doesn't auto-fix the formatting errors, but I would prefer if it did that using the `-w` argument so there's less manual work involved. It's a blocking issue anyways so let's standardize on requiring the formatting dictated by the tooling.

gfreeman-navapbc commented:
> @gfreeman-navapbc A couple notes from testing the go-import functionality:
> 
>     1. it requires a manual install:
> 
> 
> ```
> error: command not found: goimports
> ```
> 
> I had to run `go install golang.org/x/tools/cmd/goimports@latest`. Could we document that?
> 
> We could alternatively just rely on the basic `go-fmt` functionality that comes with Go, but if goimports provides some extra tooling I'd prefer to use it I think. It seems to come directly from the Golang team so I don't mind requiring it. cc. @laurenkrugen-navapbc @karinamzalez
> 
>     2. It doesn't auto-fix the formatting errors, but I would prefer if it did that using the `-w` argument so there's less manual work involved. It's a blocking issue anyways so let's standardize on requiring the formatting dictated by the tooling.

I've added the auto format to goimports. Also, judging by the documentation, it seems goimports keeps the functionality of gofmt, (including `-w`), but adds import sorting. I've also removed the old `ops/pre-commit` file entirely, I am guessing any remaining references to it are local only, as I haven't had any of the old secret check output.

kyeah reviewed (approved):
Works for me, thanks!
